File: rpn-calculator.js
// 逆ポーランド記法電卓 (手続き型スタイル)

(function () {
  "use strict";

  var _ = typeof require !== "undefined" ? require("underscore") : this._;

  var OPERATORS = {
    "+": function (x, y) {
      return y + x;
    },
    "-": function (x, y) {
      return y - x;
    },
    "*": function (x, y) {
      return y * x;
    },
    "/": function (x, y) {
      return y / x;
    }
  };

  var calculate = function (rpn) {
    var expr = rpn.split(/\s+/);
    var stack = [];
    var len = expr.length;
    var i;
    var elem;
    var res;

    for (i = 0; i < len; i += 1) {
      elem = expr[i];
      if (/^\d+$/.test(elem)) {
        stack.push(parseInt(elem, 10));
      } else {
        res = opeFn(elem)(stack.pop(), stack.pop());
        stack.push(res);
      }
    }
    return stack.pop();
  };

  var opeFn = function (ope) {
    if (_.has(OPERATORS, ope)) {
      return OPERATORS[ope];
    } else {
      throw new Error("unsupported operator '" + ope + "' is used");
    }
  };

  var RPNCalculator = {
    calculate: calculate
  };
  if (typeof module !== "undefined" && module.exports) {
    module.exports = RPNCalculator;
  } else {
    this.RPNCalculator = RPNCalculator;
  }

  // 利用例
  var rpn = "1 2 + 3 / 4 - 5 *";
  console.log(rpn, "\n = ", RPNCalculator.calculate(rpn));
}.call(this));

File: RPNCalculator.java
// 逆ポーランド記法電卓 (手続き型スタイル)

import java.util.ArrayDeque;
import java.util.Deque;

import org.apache.commons.lang3.math.NumberUtils;

class RPNCalculator {
  public static double calculate(String rpn) {
    String[] expr = rpn.split("\\s+");
    Deque<Double> stack = new ArrayDeque<>();
    for (String elem : expr) {
      if (NumberUtils.isDigits(elem)) {
        stack.addFirst(Double.parseDouble(elem));
      } else {
        double res = opeFn(elem, stack.removeFirst(), stack.removeFirst());
        stack.addFirst(res);
      }
    }
    return stack.removeFirst();
  }

  private static double opeFn(String ope, double x, double y) {
    switch (ope) {
      case "+":
        return y + x;
      case "-":
        return y - x;
      case "*":
        return y * x;
      case "/":
        return y / x;
      default:
        throw new IllegalArgumentException(
          String.format("unsupported operator '%s' is used", ope));
    }
  }

  public static void main(String[] args) {
    // 利用例
    String rpn = "1 2 + 3 / 4 - 5 *";
    System.out.printf("%s%n = %s%n", rpn, calculate(rpn));
  }
}
